rumqttc/src/lib.rs

```diff
@@ -624,7 +613,21 @@ impl MqttOptions {
     /// When set `false`, broker will hold the client state and performs pending
     /// operations on the client when reconnection with same `client_id`
     /// happens. Local queue state is also held to retransmit packets after reconnection.
+    ///
+    /// # Panic
+    ///
+    /// Panics if `clean_session` is false when `client_id` is empty.
+    ///
+    /// ```should_panic
+    /// # use rumqttc::MqttOptions;
+    /// let mut options = MqttOptions::new("", "localhost", 1883);
+    /// options.set_clean_session(false);
+    /// ```
     pub fn set_clean_session(&mut self, clean_session: bool) -> &mut Self {
+        assert!(
+            self.client_id.is_empty() && clean_session,
+            "Cannot unset clean session when client id is empty"
+        );
```

Comment:
this condition is incorrect! as this will panic even when client id isn't empty!

please consider using assert condition provided in review comment: https://github.com/bytebeamio/rumqtt/pull/791#discussion_r1475840369

you can verify it while adding tests :)